include/seqan3/io/alignment_file/format_sam_base.hpp

```diff
@@ -371,11 +371,23 @@ inline void format_sam_base::read_field(stream_view_type && stream_view,
 template <typename stream_view_type, std::ranges::forward_range target_range_type>
 inline void format_sam_base::read_field(stream_view_type && stream_view, target_range_type & target)
 {
-    if (!is_char<'*'>(*std::ranges::begin(stream_view)))
-        std::ranges::copy(stream_view | views::char_to<std::ranges::range_value_t<target_range_type>>,
-                          std::cpp20::back_inserter(target));
-    else
-        std::ranges::next(std::ranges::begin(stream_view)); // skip '*'
+    using val_t = std::ranges::range_value_t<target_range_type>;
+
+    if (auto it = std::ranges::begin(stream_view); it != std::ranges::end(stream_view))
+    {
+        char c = *it;
+        ++it;
+
+        if (it != std::ranges::end(stream_view)) // This is a string with more than one character
+        {
+            target.push_back(assign_char_to(c, val_t{}));
+            std::ranges::copy(stream_view | views::char_to<val_t>, std::cpp20::back_inserter(target));
+        }
+        else if (c != '*') // single '*' means "empty sequence"; other character is 1-element range
+        {
+            target.push_back(assign_char_to(c, val_t{}));
+        }
```

Comment:
```suggestion
        // Write to target if field does not represent an empty string, denoted as single '*' character.
        if (char c = *it; !(c == '*' && ++it == std::ranges::end(view)))
        {
            target.push_back(seqan3::assign_char_to(c, val_t{}));
            std::ranges::copy(stream_view | views::char_to<val_t>, std::cpp20::back_inserter(target));
        }
```

Reply:
This will not work for all cases, because -- as discussed above -- you are calling `std::ranges::begin()` a second time implicitly in line 382 (when iterating over `stream_view`). In the case of an underlying view that is consuming and it was at end, this will read "past-the-end".

Another issue with this code is, that in the case where `c != '*'` you are not incrementing the iterator (because of lazy-evaluation of `&&`) and thus writing `c` to the target twice.

I don't like have a line twice in the code, but I think it is the easiest (correct) solution.

Reply:
>This will not work for all cases, because -- as discussed above -- you are calling std::ranges::begin() a second time implicitly in line 382 (when iterating over stream_view). In the case of an underlying view that is consuming and it was at end, this will read "past-the-end".

Oh, that is actually true. Good point. But nothing that can be easily fixed by calling the right interface:
Since we already have the iterator we can just use it to form a subrange which does not consume anything
```cpp
std::ranges::copy(std::ranges::subrange<...>{it, std::ranges::end(view)} | views::char_to<val_t>, ...);
```
That should work right?

> Another issue with this code is, that in the case where c != '*' you are not incrementing the iterator (because of lazy-evaluation of &&) and thus writing c to the target twice.

Yeah, that's right. I had the order the other way around, which schould be guaranteed to work. In the end, I thought it is nicer to have the char check before but this would introduce a subtle bug indeed.

Reply:
> That should work right?

Yes, the proposed solution should work and is more elegant! Although I think it is considerably more difficult to parse for humans while the slightly longer one is a lot more straight-forward (just thinking about how easy it was for you to make this mistake and that I needed to think hard on whether it works now 🤓 ).

But I will leave others to judge 😉 I have no strong preference about it.

Reply:
The problem is that when writing an algorithm with ranges you should pay attention to when you call begin. I don't think  that more verbose code makes the actual problem clearer, as it is quite subtle. At least not for someone who is reading the code for the first time. So I would support the concise version with a clearer comment at the beginning.

Reply:
Anyhow, the introduced issues make up some good test cases that should be added.

Reply:
> Anyhow, the introduced issues make up some good test cases that should be added.

I don't really see the good test cases. The function in question is already used with a consuming and a non-consuming input range, both empty (`*`) in some cases and non-empty in others. The case of a field starting with `*` but not being empty was also added (I added additionally added for the id field which is consuming in the test case - that one is new).